File: backend/src/services/apolloScraper.js
/**
 * Apollo.io API Integration
 *
 * Uses Apollo's free-tier "People Search" and "Organization Search" endpoints
 * to find companies and contacts matching the target config.
 *
 * Free tier: 50 credits/month. Each search uses ~1 credit per result.
 * API docs: https://apolloio.github.io/apollo-api-docs/
 */

const APOLLO_BASE = 'https://api.apollo.io/v1';

export async function searchCompanies(config, apiKey) {
  if (!apiKey) {
    console.warn('Apollo API key not set — skipping Apollo search');
    return [];
  }

  const {
    industry = '',
    location = '',
    companySize = '',
    roleKeywords = '',
  } = config;

  try {
    // Step 1: Search for organizations matching filters
    const orgRes = await fetch(`${APOLLO_BASE}/mixed_companies/search`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Cache-Control': 'no-cache',
      },
      body: JSON.stringify({
        api_key: apiKey,
        q_organization_keyword_tags: industry ? [industry] : undefined,
        organization_locations: location ? [location] : undefined,
        organization_num_employees_ranges: companySize
          ? [companySize]
          : undefined,
        page: 1,
        per_page: 25, // conservative to stay within free tier
      }),
    });

    if (!orgRes.ok) {
      const err = await orgRes.text();
      throw new Error(`Apollo org search failed (${orgRes.status}): ${err}`);
    }

    const orgData = await orgRes.json();
    const organizations = orgData.organizations || [];

    if (organizations.length === 0) {
      console.log('Apollo: no organizations found for given filters');
      return [];
    }

    // Step 2: For each org, find a relevant contact (hiring manager / founder)
    const results = [];

    for (const org of organizations) {
      try {
        const contact = await findContact(org.id, roleKeywords, apiKey);

        results.push({
          name: org.name,
          website: org.website_url || org.primary_domain,
          contactName: contact?.name || null,
          contactEmail: contact?.email || null,
          contactTitle: contact?.title || null,
          linkedinUrl: org.linkedin_url || null,
          industry: org.industry || null,
          location: org.city
            ? `${org.city}, ${org.state || org.country}`
            : org.country || null,
          size: org.estimated_num_employees
            ? estimateRange(org.estimated_num_employees)
            : null,
          techStack: org.technology_names?.slice(0, 10) || [],
          recentNews: null, // enriched later via web search
          roleHint: contact?.title || roleKeywords || null,
          source: 'apollo',
        });
      } catch (err) {
        console.error(`Apollo: error fetching contact for ${org.name}:`, err.message);
        // Still include the company without contact info
        results.push({
          name: org.name,
          website: org.website_url || org.primary_domain,
          contactName: null,
          contactEmail: null,
          contactTitle: null,
          linkedinUrl: org.linkedin_url || null,
          industry: org.industry || null,
          location: org.city
            ? `${org.city}, ${org.state || org.country}`
            : org.country || null,
          size: org.estimated_num_employees
            ? estimateRange(org.estimated_num_employees)
            : null,
          techStack: org.technology_names?.slice(0, 10) || [],
          recentNews: null,
          roleHint: roleKeywords || null,
          source: 'apollo',
        });
      }
    }

    return results;
  } catch (err) {
    console.error('Apollo search error:', err.message);
    return [];
  }
}

/**
 * Find the most relevant contact at a company (HR, founder, engineering lead).
 * Uses Apollo's people search scoped to a specific organization.
 */
async function findContact(orgId, roleKeywords, apiKey) {
  // Search for people with relevant titles at this org
  const titles = roleKeywords
    ? [roleKeywords]
    : ['HR', 'Hiring Manager', 'Engineering Manager', 'CTO', 'Founder'];

  const res = await fetch(`${APOLLO_BASE}/mixed_people/search`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Cache-Control': 'no-cache',
    },
    body: JSON.stringify({
      api_key: apiKey,
      q_organization_id: orgId,
      person_titles: titles,
      page: 1,
      per_page: 1,
    }),
  });

  if (!res.ok) return null;

  const data = await res.json();
  const person = data.people?.[0];
  if (!person) return null;

  return {
    name: person.name,
    email: person.email,
    title: person.title,
    linkedin: person.linkedin_url,
  };
}

// Converts a raw employee count to a range string (e.g. "51-200")
function estimateRange(count) {
  if (count <= 10) return '1-10';
  if (count <= 50) return '11-50';
  if (count <= 200) return '51-200';
  if (count <= 1000) return '201-1000';
  return '1000+';
}

File: app/services/apollo_scraper.py
from __future__ import annotations
"""
Apollo.io API Integration.

Free tier: 50 credits/month. Each search uses ~1 credit per result.
Searches for companies + contacts matching target config.
"""

import httpx
from app.config import get_settings

APOLLO_BASE = "https://api.apollo.io/v1"


async def search_companies(config: dict) -> list[dict]:
    """Search Apollo for companies matching filters. Returns list of company dicts."""
    api_key = get_settings().apollo_api_key
    if not api_key:
        print("Apollo API key not set — skipping")
        return []

    industry = config.get("industry", "")
    location = config.get("location", "")
    company_size = config.get("companySize", "")
    role_keywords = config.get("roleKeywords", "")

    try:
        async with httpx.AsyncClient(timeout=30) as client:
            # Step 1: Search organizations
            payload = {
                "api_key": api_key,
                "page": 1,
                "per_page": 25,
            }
            if industry:
                payload["q_organization_keyword_tags"] = [industry]
            if location:
                payload["organization_locations"] = [location]
            if company_size:
                payload["organization_num_employees_ranges"] = [company_size]

            resp = await client.post(f"{APOLLO_BASE}/mixed_companies/search", json=payload)
            resp.raise_for_status()
            orgs = resp.json().get("organizations", [])

            if not orgs:
                print("Apollo: no organizations found")
                return []

            # Step 2: For each org, find a relevant contact
            results = []
            for org in orgs:
                contact = await _find_contact(client, org["id"], role_keywords, api_key)
                results.append({
                    "name": org.get("name"),
                    "website": org.get("website_url") or org.get("primary_domain"),
                    "contactName": contact.get("name") if contact else None,
                    "contactEmail": contact.get("email") if contact else None,
                    "contactTitle": contact.get("title") if contact else None,
                    "linkedinUrl": org.get("linkedin_url"),
                    "industry": org.get("industry"),
                    "location": _build_location(org),
                    "size": _estimate_range(org.get("estimated_num_employees")),
                    "techStack": (org.get("technology_names") or [])[:10],
                    "recentNews": None,
                    "roleHint": (contact.get("title") if contact else None) or role_keywords or None,
                    "source": "apollo",
                })

            return results

    except Exception as e:
        print(f"Apollo search error: {e}")
        return []


async def _find_contact(client: httpx.AsyncClient, org_id: str, role_keywords: str, api_key: str) -> dict | None:
    """Find a relevant contact (HR/founder/eng lead) at a company."""
    titles = [role_keywords] if role_keywords else ["HR", "Hiring Manager", "Engineering Manager", "CTO", "Founder"]

    try:
        resp = await client.post(f"{APOLLO_BASE}/mixed_people/search", json={
            "api_key": api_key,
            "q_organization_id": org_id,
            "person_titles": titles,
            "page": 1,
            "per_page": 1,
        })
        if resp.status_code != 200:
            return None

        person = resp.json().get("people", [None])[0]
        if not person:
            return None

        return {"name": person.get("name"), "email": person.get("email"), "title": person.get("title")}
    except Exception:
        return None


def _build_location(org: dict) -> str | None:
    city = org.get("city")
    state = org.get("state")
    country = org.get("country")
    if city:
        return f"{city}, {state or country}"
    return country


def _estimate_range(count: int | None) -> str | None:
    if not count:
        return None
    if count <= 10:
        return "1-10"
    if count <= 50:
        return "11-50"
    if count <= 200:
        return "51-200"
    if count <= 1000:
        return "201-1000"
    return "1000+"
